### DISCUSSION TITLE
Tool Registry

### DISCUSSION THREAD

**Discussion by aboy021** [babashka/babashka#1018] [Ideas]
There are all kinds of useful tools that you can put together using Babashka and pods. It might be nice if there was a way of publishing tools to a registry that other users could install/update/remove. For example, the [http-server.clj](https://github.com/babashka/babashka/blob/master/examples/http-server.clj) example would be great as a tool.

Some features that could be included:

- Shims for installed tools so that you can call them directly (e.g. http-server)
- Aliasing so that you can change the name of the tool's shim and avoid collisions
- Listing of installed tools
- Support for multiple tool registries (public and private)
- Autocomplete support

It might be worth implementing this as a stand alone tool as Babashka already has a lot of functionality.

**Comment by aboy021** [comment]
Doing some more reading this would tie into the discussion about global tasks:

- [Global tasks: TL;DR A shell alias (or global script) is probably the better answer to global tasks. · Issue #819 · babashka/babashka](https://github.com/babashka/babashka/issues/819)

**Comment by borkdude** [comment]
> It might be worth implementing this as a stand alone tool as Babashka already has a lot of functionality.

I think it would be convenient to have this in babashka (`bb install foo/bar --alias bar2 && bar2`) but this can probably be prototyped using a user-space babashka script first. I think it's a neat idea that I've been thinking about myself for a while as well.

One such tool I have myself is https://github.com/borkdude/gh-release-artifact which could be `gh-release ... ...` on the command line if you installed it. 

If we installed shims in `~/.babashka/tools/gh-release`, `~/.babashka/tools/http-server` then one would only have to put `~/.babashka/tools` on the `PATH` and autocomplete would already work. The shims can be bash or cmd (Windows) shims that just call babashka, or babashka scripts directly (`#!/usr/bin/env bb`, for Windows there is a more complex but also working one). One benefit of having bash/cmd shims is that you can `cd` into another directory and then execute whatever bb script/task.

**Comment by borkdude** [comment]
Just to list more options: re-use existing package managers for these tools.

`brew install babashka/tools/http-server` which would have a dependency on bb, would git clone the project and invoke `bb http-server` from the cloned repo.

**Comment by borkdude** [comment]
I have made a prototype for installing a babashka script as a brew command here:

https://gist.github.com/borkdude/1a53bfff44fa3d4f262148e9e995a226
